File: aoc2024/src/python/day3.py
from operator import mul
import time
from aoc_utils import runIt, PuzzleInput
import functools
import math
import re
import itertools
import sys


testInput = r"""xmul(2,4)%&mul[3,7]!@^do_not_mul(5,5)+mul(32,64]then(mul(11,8)mul(8,5))"""
input = PuzzleInput('input/day3.txt', testInput)

lines = input.getInputLines(test=False)
line = ''.join(lines)
p = re.compile(r'mul\((\d+),(\d+)\)')

def part1():
    c = sum(map(lambda expr: mul(*map(int, expr.group(1, 2))), p.finditer(line)))
    print(f'sum {c}')

def part2():
    enabled = True
    c = 0
    for mulMatch in p.finditer(line):
        lastDo = line[:mulMatch.span()[0]].rfind('do()')
        lastDont = line[:mulMatch.span()[0]].rfind('don\'t()')
        if (lastDont > lastDo):
            enabled = False
        else:
            enabled = True

        if enabled:
            c += mul(*map(int, mulMatch.group(1, 2)))

    print(f'sum {c}')


runIt(part1, part2)

File: aoc2024/src/ts/day3.ts
import {readInput} from './utils/file-utils';

let lines = readInput('day3', false);


function part1() {
    const mulRegex = /mul\((\d+),(\d+)\)/gd

    var sum = BigInt(0)
    lines.forEach(line => {        
        const matches = line.matchAll(mulRegex);

        if (matches) {
            for (let match of matches) {
                const index = match.indices
                const product = parseInt(match[1]) * parseInt(match[2]);

                // console.log(`match: ${match[0]} - ${match[1]} ${match[2]} = ${product}`)
                sum = sum + BigInt(product)
            };
        } else {
            console.log('No mul found')
        }
    });
    console.log(`sum: ${sum}`)
}

function part2() {
    const mulRegex = /mul\((\d+),(\d+)\)/gd
    const doRegex = /do\(\)/gd
    const dontRegex = /don\'t\(\)/gd


    var sum = BigInt(0)
    var enabled = true;

    lines.forEach(line => {        
        const matches = line.matchAll(mulRegex);
        var matchIndex = 0;

        var index = 0;
        if (matches) {
            for (let match of matches) {
                // look for any dos or donts before the next match, last one wins.
                const nextMatchIndex = match.indices!![0][0]

                var substr = line.substring(matchIndex, nextMatchIndex);
                var lastDo = substr.lastIndexOf('do()');
                var lastDont = substr.lastIndexOf('don\'t()');
                if (lastDo > -1 && lastDo > lastDont) {
                    enabled = true;
                } else if (lastDont > -1 && lastDont > lastDo) {
                    enabled = false;
                }

                if (enabled) {
                    const product = parseInt(match[1]) * parseInt(match[2]);

                    // console.log(`match: ${match[0]} - ${match[1]} ${match[2]} = ${product}`)
                    sum = sum + BigInt(product)
                }
                matchIndex = nextMatchIndex;
            };
        } else {
            console.log('No mul found')
        }
    });
    console.log(`sum: ${sum}`)
}

part2();
